--- a/src/third_party/WebKit/WebCore/editing/TextIterator.cpp
+++ b/src/third_party/WebKit/WebCore/editing/TextIterator.cpp
@@ -1740,6 +1740,11 @@ inline SearchBuffer::SearchBuffer(const
 
 inline SearchBuffer::~SearchBuffer()
 {
+    // Leave the static object pointing to a valid string.
+    UErrorCode status = U_ZERO_ERROR;
+    usearch_setPattern(WebCore::searcher(), &newlineCharacter, 1, &status);
+    ASSERT(status == U_ZERO_ERROR);
+
     unlockSearcher();
 }
 
