--- a/src/third_party/WebKit/WebCore/rendering/RenderLineBoxList.cpp
+++ b/src/third_party/WebKit/WebCore/rendering/RenderLineBoxList.cpp
@@ -322,11 +322,9 @@ void RenderLineBoxList::dirtyLinesFromCh
         adjacentBox = box->prevRootBox();
         if (adjacentBox)
             adjacentBox->markDirty();
-        if (child->isBR() || (curr && curr->isBR())) {
-            adjacentBox = box->nextRootBox();
-            if (adjacentBox)
-                adjacentBox->markDirty();
-        }
+        adjacentBox = box->nextRootBox();
+        if (adjacentBox && (adjacentBox->lineBreakObj() == child || child->isBR() || (curr && curr->isBR())))
+            adjacentBox->markDirty();
     }
 }
 
