--- a/src/third_party/WebKit/WebCore/rendering/RenderBlock.cpp
+++ b/src/third_party/WebKit/WebCore/rendering/RenderBlock.cpp
@@ -2829,7 +2829,7 @@ void RenderBlock::removeFloatingObject(R
                     // Special-case zero- and less-than-zero-height floats: those don't touch
                     // the line that they're on, but it still needs to be dirtied. This is
                     // accomplished by pretending they have a height of 1.
-                    bottom = max(bottom, it.current()->m_top + 1);
+		    bottom = max(bottom, it.current()->m_top == numeric_limits<int>::max() ? it.current()->m_top : it.current()->m_top + 1);
                     markLinesDirtyInVerticalRange(0, bottom);
                 }
                 m_floatingObjects->removeRef(it.current());
@@ -3415,7 +3415,7 @@ void RenderBlock::markLinesDirtyInVertic
 
     RootInlineBox* lowestDirtyLine = lastRootBox();
     RootInlineBox* afterLowest = lowestDirtyLine;
-    while (lowestDirtyLine && lowestDirtyLine->blockHeight() >= bottom) {
+    while (lowestDirtyLine && lowestDirtyLine->blockHeight() >= bottom && bottom < numeric_limits<int>::max()) {
         afterLowest = lowestDirtyLine;
         lowestDirtyLine = lowestDirtyLine->prevRootBox();
     }
