--- a/src/third_party/WebKit/WebCore/svg/SVGFont.cpp
+++ b/src/third_party/WebKit/WebCore/svg/SVGFont.cpp
@@ -242,7 +242,7 @@ struct SVGTextRunWalker {
     {
         ASSERT(0 <= from && from <= to && to - from <= run.length());
 
-        const String text = Font::normalizeSpaces(String(run.data(from), run.length()));
+        const String text = Font::normalizeSpaces(String(run.data(from), to - from));
         Vector<SVGGlyphIdentifier::ArabicForm> chars(charactersWithArabicForm(text, run.rtl()));
 
         SVGGlyphIdentifier identifier;
