--- a/src/third_party/WebKit/WebCore/dom/StyleElement.cpp
+++ b/src/third_party/WebKit/WebCore/dom/StyleElement.cpp
@@ -64,8 +64,12 @@ void StyleElement::process(Element* e, i
     unsigned resultLength = 0;
     for (Node* c = e->firstChild(); c; c = c->nextSibling()) {
         Node::NodeType nodeType = c->nodeType();
-        if (nodeType == Node::TEXT_NODE || nodeType == Node::CDATA_SECTION_NODE || nodeType == Node::COMMENT_NODE)
-            resultLength += c->nodeValue().length();
+        if (nodeType == Node::TEXT_NODE || nodeType == Node::CDATA_SECTION_NODE || nodeType == Node::COMMENT_NODE) {
+	    unsigned length = c->nodeValue().length();
+	    if (length > std::numeric_limits<unsigned>::max() - resultLength)
+		CRASH();
+	    resultLength += length;
+	}
     }
     UChar* text;
     String sheetText = String::createUninitialized(resultLength, text);
